### DISCUSSION TITLE
fmt format

### DISCUSSION THREAD

**Discussion by xphoenix** [foonathan/lexy#217] [Q&A] [answered]
Good day,

Im trying to write grammar for `fmt::format`. The main problem im struggling with is how to match "anything outside of {...}".
```
struct grammar_format_string {
    static constexpr auto rule = [] {
        auto param_start = dsl::not_followed_by(dsl::lit_c<'{'>, dsl::lit_c<'{'>);
        return 
              // That would be "text before param if any" then param if any
              // However it doesn't work as param_start consumes first '{'
              //
              // Even if make it works and wrap in a kind of list/loop = it seems to be not the most
              // performance wise way of doing a such parsing
              dsl::until(param_start) + dsl::opt(grammar_parameter);
    }();
};
```

Please guide me in which direction to look

**Comment by foonathan** [comment, marked as answer]
Logically, you have a list of segments, where each segment is either a unicode code point, or a `{...}` sequence:

```cpp
dsl::list(dsl::curly_bracketed(format_syntax) | dsl::else_ >> dsl::code_point)
```

(use `dsl::bracekts(dsl::not_followed_by(...), ...)` as appropriate)

Alternatively, you can merge adjacent unescaped code points, you can use `dsl::until`, which will consume the `{`, but this just means you don't have to handle it specially. I think you could just treat it as a list of segments, with `{` serving as list separator:

```cpp
dsl::list(dsl::until(param_start).or_eof(), dsl::sep(param_start >> grammar_parameter + dsl::lit_c<'}'>))
```
